### DISCUSSION TITLE
联机灵感

### DISCUSSION THREAD

**Discussion by wudiwudidream** [Meloong-Git/PCL#4158] [提问] [answered]
快速开发联机最重要的是直接套用框架，可套用框架Radmin VPN软件的联机框架

**Comment by MoYuan-CN** [comment, marked as answer]
参见 #2128 #3899 等

**Reply by wudiwudidream** [reply]
Radmin VPN这个用过，比樱花稳

**Reply by MoYuan-CN** [reply]
PCL 不会随意内置其他软件的……

**Reply by wudiwudidream** [reply]
我说的是套用框架

**Reply by allMagicNB** [reply]
> 我说的是套用框架

不管怎样，你在想屁吃。

**Reply by wudiwudidream** [reply]
我气笑了，复制粘贴不会啊

**Reply by allMagicNB** [reply]
@LTCatt 请求 Ban：@wudiwudidream

**Reply by MoYuan-CN** [reply]
打，打，打住
无论如何，别在 Disucssion 楼内吵架

**Reply by wudiwudidream** [reply]
OK

**Reply by wudiwudidream** [reply]
@allMagicNB 好家伙，又没违反@LTCatt 设立的规则（笑不活）

**Reply by LTCatt** [reply]
楼主可以注意一下，在发帖之前先搜搜有没有重复内容。
@allMagicNB 首先他也没违规，其次先开始不友善发言的是你，我觉得你的问题更大一点。你对 “看起来不怎么熟悉规矩的新人” 开门见山地不友好发言已经不是一次两次了，**请对新人友好一些。**

**Reply by wudiwudidream** [reply]
@LTCatt 龙猫腾跃，启动器开发就属你最靠谱，我谈的是Python socket 那个库实验下来是可以完成联机的

**Reply by allMagicNB** [reply]
¯\\_(ツ)\_/¯

**Reply by wudiwudidream** [reply]
@LTCatt 我的想法是：仿照Minecraft局域网联机将收集传输信息，将主机设为服务器，通过socket传输json文件或其他文件直接传输（java同样可以做到，唯一担心的点是延迟不稳定问题，根据Among us三小时掉线81次的经验，不能直接照仿官方给的局域网联机），内网穿透，剩下是按照标准的TCP协议和计算机网络运行过程

**Reply by LTCatt** [reply]
多谢提供信息，不过目前联机并不是遇到了技术问题。等到网易不代理国服了联机就有了（

**Reply by wudiwudidream** [reply]
@LTCatt  那个简单，1、没有违反中国和美国的相关规定，毕竟是国际版  2、网易是代理商，国际版联机是微软说的算而不是网易说的算。3、没有违反网易禁止私人服务器规定（虽说这个搞得跟没说一样，因为不是微软规定的，所以国际版私人服务器影响不到） 4、没有违反[Minecraft 最终用户许可协议](https://account.mojang.com/documents/minecraft_eula)和[隐私政策](https://go.microsoft.com/fwlink/?LinkId=521839)。，和EULA协议就可以

**Reply by wudiwudidream** [reply]
有一个点：上一次你添加联机功能的时候，有考虑到后面提供联机厂商搞出后面的情况？（联机别照仿需要服务器作为中间的那种方式），也是不推荐的，况且那时候已经支持不用中间服务器直接传输到其他主机的技术，不用内网穿透的，仿照网站编写模式就可以编写不用内网穿透的联机方式，况且Django等这种库自带链接自定义，解决输入IPV4/6过长输入慢的问题，而且Among us每次需要验证用户档案导致延迟过高等问题你觉得你解决了？

**Reply by MoYuan-CN** [reply]
> https://github.com/Hex-Dragon/PCL2/discussions/4158#discussioncomment-9968228
> https://github.com/Hex-Dragon/PCL2/discussions/4158#discussioncomment-9968235
> https://github.com/Hex-Dragon/PCL2/discussions/4158#discussioncomment-9968236

我们能理解您对联机功能的需求之高，但是也请不要刷屏，谢谢

> 况且Django等这种库自带链接自定义

Django 为 Python 库，PCL 启动器是用 VB 开发，代码语言不同可能会带来额外的开发成本

> 已经支持不用中间服务器直接传输到其他主机的技术，不用内网穿透的

Octo Network（BakaXL 内测中）= P2P 网络

> 仿照网站编写模式就可以编写不用内网穿透的联机方式

这一点我比较感兴趣，如果可以的话作者能联系我交流交流嘛？

**Reply by wudiwudidream** [reply]
OK

**Reply by wudiwudidream** [reply]
不是刷屏，我没用VPN延迟很高，卡出来的

**Reply by wuliaodexiaoluo** [reply]
> 有一个点：上一次你添加联机功能的时候，有考虑到后面提供联机厂商搞出后面的情况？（联机别照仿需要服务器作为中间的那种方式），也是不推荐的，况且那时候已经支持不用中间服务器直接传输到其他主机的技术，不用内网穿透的，仿照网站编写模式就可以编写不用内网穿透的联机方式，况且Django等这种库自带链接自定义，解决输入IPV4/6过长输入慢的问题，而且Among us每次需要验证用户档案导致延迟过高等问题你觉得你解决了？

那没办法打洞的 NAT 4 不还是需要中转服务器

>3、没有违反网易禁止私人服务器规定（虽说这个搞得跟没说一样，因为不是微软规定的，所以国际版私人服务器影响不到）

网易懒得搞你而已，要是真搞起来个人不一定搞得过大公司，看看 HiPer 和 TGZ 事件就知道龙猫可能并不愿意处理这种纠纷
人家没权管私服，但是不影响人家起诉你，强行给你扣个修改游戏的帽子

>套用框架

套框架写代码的时间估计足够把所有 Issue 写完了

**Comment by wudiwudidream** [comment]
***@***.***的自动回复邮件。

**Comment by RimacC2-EV** [comment]
XDM我有几个<del>不知天高地厚的</del>问题，能请教各路大佬吗：

我记得 ToDesk 好像有个“端到端连接”，听起来似乎与PCL2的联机方式有那么一丢丢的联系？


还有，能不能用类似于 P2P/迅雷网络 的方式联机？

**Reply by allMagicNB** [reply]
ToDesk 是端到端加密，实际上还是连接到了中转服务器。
P2P 还没时间做……
迅雷网络？哥们你是真不知道迅雷背后那丑恶的嘴脸。

**Reply by RimacC2-EV** [reply]
好吧······

端到端加密的 中转服务器 能不能招募志愿者？<del>能拉我当小白鼠去吗  虽然我凌晨基本不在线</del>

**Reply by Pigeon0v0** [reply]
Octo Network 就是使用的 P2P 技术，加入者会和主机尝试建立点对点连接，仅在直接连接失败且双方同意进行中继时才会通过中继连接

你先别急.jpg

**Comment by wudiwudidream** [comment]
@Pigeon0v0 我也有，试问：内网穿透是不是根据TCP协议，并且是使用各个语言统一代称socket库编写？

**Reply by wuliaodexiaoluo** [reply]
实际上 UDP 也可以，只要事先商定好发包格式

**Comment by Silverteal** [comment]
如果缺乏相关知识基础，可以寻找更适合的地方进行学习，这里并不适合。
